### ISSUE TITLE
DataTable sort not always "stable"

### ISSUE THREAD

**Issue by tdennist** [bokeh/bokeh#1700]
The sort appears not always to be a stable sort.  Stable sorting is actually quite important in a table widget as it is quite common to display data where unique values are defined by the combination of multiple columns.  Right now the order appears to be random sometimes.  Is this a feature of the underlying table library SlickGrid?  I googled around a bit and one of the SlickGrid discussions suggested that the sort was stable but it was a little unclear: https://github.com/mleibman/SlickGrid/issues/9

To reproduce the problem run the following code in an ipython notebook and then click on column A, then column B.  If you don't see the problem try A and then B again.  After 1,2 or 3 tries it almost always happens.

```
from bokeh import models, plotting, widgets

j = 10
df = pandas.DataFrame(dict(A=['A']*j + ['B']*j, B=[i for i in range(j)]*2))
source = models.ColumnDataSource(df)
columns = [
    widgets.TableColumn(field=c, title=c) for c in df.columns
]
data_table = widgets.DataTable(source=source, columns=columns, editable=True, width=1000)
plotting.show(data_table)
```

I'm using bokeh '0.7.1.dev.20150109' and Chrome Version 39.0.2171.71

**Comment by bryevdv** [comment]
I believe this was fixed by some of the recent DataTable work for `0.12.6`. I am unable to reproduce the problem any more with the code above, even after sorting alternate columns many dozens of times. Please re-open if this can be reproduced in recent Bokeh versions.

**Comment by nicain** [comment]
Hi All,

Unfortunately I am still seeing this bug with bokeh.__version__ = 0.12.6rc3:

from bokeh import models, plotting
from bokeh.models import widgets
import pandas

df = pandas.DataFrame(dict(col1=['A','A','A','B','B','B', 'C', 'C', 'C'],
                           col2=[1,1,2,2,2,3,3,3,4]))
source = models.ColumnDataSource(df)
columns = [
    widgets.TableColumn(field=c, title=c) for c in df.columns
]
data_table = widgets.DataTable(source=source, columns=columns, editable=True, width=1000)
plotting.show(data_table)

I reliably reproduce this by sorting by column "col2" twice; afterwards, if you look in column "col1", the A's and C's are not still grouped together, which I think should be the result of a stable sort.

**Comment by bryevdv** [comment]
Ah I was only checking that the full return was in the same order, sorry.

**Comment by bryevdv** [comment]
So, I replaced `Array.sort` with a simple merge sort implementation, which must be stable, and I get the same results. Which got me thinking that perhaps it is correct as is. 

Let's take a look at things. Here is a screenshot of the original table initially:

<img width="675" alt="screen shot 2017-06-17 at 23 31 21" src="https://user-images.githubusercontent.com/1078448/27258053-1c8e32b8-53b5-11e7-8984-2bd13a687f7b.png">

And here is a screenshot of the table after sorting the second column descending:

<img width="634" alt="screen shot 2017-06-17 at 23 31 26" src="https://user-images.githubusercontent.com/1078448/27258054-20b89914-53b5-11e7-8dcd-d06d44c63978.png">

What does a stable sort guarantee? My understanding is that: elements that ***compare equal*** are maintained in the same relative order, after the sort. But that is exactly what is seen here? We have to compare the groups where `col2` is equal, because that is all a stable sort makes a claim about:

* The bottom to "A" elements compare equal, because they are `col2=1` and are in the same relative order: 0A is above 1A, just as in the first image

* The next three elements up compare equal, because they are `col2=2` and are in the same relative order: 3B is above 4B, but more important 2A is above all the B's that it can be, just as it is in the first image

* same story for the next three: 6C is above 76 (as before) and 5B is above all the C's that it can be, constrained by `col2=3`

* the last C has to be above everything, `col2=4`

My understanding of a stable sort is that it is most useful for sorting multiple columns, to keep the previous columns in the same order. And that is exactly what this this does. Here is the same example with a`col3` added. If I first sort `col2` then sort `col3` ascending and then descending, the order of `col2` stays the same for both:

<img width="1045" alt="screen shot 2017-06-17 at 23 43 39" src="https://user-images.githubusercontent.com/1078448/27258108-0121920c-53b7-11e7-9003-0c219d752be4.png">

<img width="1064" alt="screen shot 2017-06-17 at 23 43 44" src="https://user-images.githubusercontent.com/1078448/27258109-04b3abe4-53b7-11e7-93ac-728ea0450c01.png">

@nicain do you have a different understanding or definition of stable sort that contradicts this analysis? I'm happy to submit the PR with the merge sort, because I do believe some browsers do not implement a stable sort by default, so this would make things uniform. (I think safari does, and the unchanged results are a coincidence)

**Comment by bryevdv** [comment]
I should add, running also the original code, the issue initially stated that things would be "random" after sorting a few times. That's what I might expect from a truly non-stable sort. But if I run that code now and sort back and for I always get the same orders, which is what I would expect from a stable sort. 

@nicain perhaps the issue is that you are assuming that `col1` is implicitly sorted? It's definitely not! if you click on `col2` then ordering by `col2` is the only criteria, and stable sorting seems to be satisfied by my observations above. But if I first sort `col1` and then shift click to secondarily sort `col2` then I see the result I think you are expecting:

<img width="706" alt="screen shot 2017-06-17 at 23 56 20" src="https://user-images.githubusercontent.com/1078448/27258161-ce34fac6-53b8-11e7-8e1c-6c25d1259cb1.png">


but again, that's only because in this instance, I have *explicitly* sorted `col1` first.

**Comment by nicain** [comment]
I dug into this a little bit, and I agree that the sort is stable; I am afraid the problem is with the user (me)!  What has been tripping me up is that the first click on a given column is always an ascending (stable, btw) sort.  When I was sorting one column descending (by clicking twice), my brain was assuming that the next column would also be a descending sort.  In reality, the second column by default starts with a ascending sort, the opposite of what I expected. I then got fooled by looking back at the first column, when I should have just clicked again on the second column sort, implementing a (stable!) descending sort on column 2.  Because column 1 was reversed, and the sort is stable, it results in my intended behavior.

To summarize, yes the sort is stable, and there is no bug! However, I would be in favor of having the default sort direction (i.e ascending or descending) be remembered from any previous sort; after all, no matter the key, sort direction is independent of column and it is a little cognitively jarring to perform two sorts and have them (because of the default setting) perform different behavior.  But this is really just what my brain expected, so if this isn't a problem for others I am fine being overruled.

**Comment by bryevdv** [comment]
After some additional thought I am going to close this. Partly because I believe  the `effort required/frequency of occurrence` is regrettably large (no one else has since mentioned this situation) but also because there is technically a sort indicator on the columns that always does show the current sort order.
